js/pro/woo.js

```diff
@@ -0,0 +1,722 @@
+        //         "size":300,
+        //         "side":"BUY",
+        //         "source":0
+        //     }
+        //
+        const marketId = this.safeString (trade, 'symbol');
+        market = this.safeMarket (marketId, market);
+        const symbol = market['symbol'];
+        const price = this.safeString (trade, 'price');
+        const amount = this.safeString (trade, 'size');
+        const cost = Precise.stringMul (price, amount);
+        const side = this.safeStringLower (trade, 'side');
+        const timestamp = this.safeInteger (trade, 'timestamp');
+        return this.safeTrade ({
+            'id': undefined,
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'symbol': symbol,
+            'side': side,
+            'price': price,
+            'amount': amount,
+            'cost': cost,
+            'order': undefined,
+            'takerOrMaker': undefined,
+            'type': undefined,
+            'fee': undefined,
+            'info': trade,
+        }, market);
+    }
+
+    checkRequiredUid (error = true) {
+        if (!this.uid) {
+            if (error) {
+                throw new AuthenticationError (this.id + ' requires `uid` credential');
+            } else {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    async authenticate (params) {
+        this.checkRequiredCredentials ();
+        const url = this.urls['api']['ws']['private'] + '/' + this.uid;
+        const client = this.client (url);
+        const messageHash = 'authenticated';
+        const event = 'auth';
+        let future = this.safeValue (client.subscriptions, messageHash);
+        if (future === undefined) {
+            future = client.future (messageHash);
+            const ts = this.nonce ().toString ();
+            let auth = this.urlencode (params);
+            auth += '|' + ts;
+            const signature = this.hmac (this.encode (auth), this.encode (this.secret), 'sha256');
+            const request = {
+                'event': event,
+                'params': {
+                    'apikey': this.apiKey,
+                    'sign': signature,
+                    'timestamp': ts,
+                },
+            };
+            this.spawn (this.watch, url, messageHash, request, messageHash, future);
+        }
+        return await future;
+    }
+
+    async watchPrivate (messageHash, message, params = {}) {
+        await this.authenticate (params);
+        const url = this.urls['api']['ws']['private'] + '/' + this.uid;
+        const requestId = this.requestId (url);
+        const subscribe = {
+            'id': requestId,
+        };
+        const request = this.extend (subscribe, message);
+        return await this.watch (url, messageHash, request, messageHash, subscribe);
+    }
+
+    async watchOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
```

Comment:
I'm also getting this signature error:
![image](https://user-images.githubusercontent.com/43336371/205096819-e722fd6f-09cf-4f49-b25f-10a318a08332.png)

Reply:
It worked for me.

$ node examples/js/cli woo watchOrders --test --verbose
Node.js: v18.8.0
CCXT v2.1.96
woo.watchOrders ()
2022-12-05T06:40:51.243Z connecting to wss://wss.staging.woo.org/v2/ws/private/stream/...
2022-12-05T06:40:51.335Z onUpgrade
2022-12-05T06:40:51.336Z onOpen
2022-12-05T06:40:51.336Z sending {
  event: 'auth',
  params: {
    apikey: '...',
    sign: '...',
    timestamp: '1670222451242'
  }
}
2022-12-05T06:40:51.392Z onMessage { event: 'auth', success: true, ts: 1670222451420 }
2022-12-05T06:40:51.492Z sending { id: 1, event: 'subscribe', topic: 'executionreport' }
2022-12-05T06:40:51.545Z onMessage { id: '1', event: 'subscribe', success: true, ts: 1670222451577 }
2022-12-05T06:40:59.976Z onMessage { event: 'ping', ts: 1670222460000 }
2022-12-05T06:41:01.337Z sending { event: 'ping' }
2022-12-05T06:41:01.391Z onMessage { event: 'pong', ts: 1670222461419 }

Reply:
@sc0Vu humm weird, will try again